src/common/core/metaquerycore.js

```diff
@@ -0,0 +1,261 @@
+                sensitiveFilter(validNodes);
+            }
+
+            //before every next step we check if we still have potential nodes
+            if (validNodes.length === 0) {
+                return validNodes;
+            }
+
+            if (parameters.multiplicity === true) {
+                if (rules.max && rules.max > -1 && innerCore.getChildrenRelids(node).length >= rules.max) {
+                    validNodes = [];
+                    return validNodes;
+                }
+                if (children.length === 0) {
+                    return validNodes; //we cannot check type-multiplicity without children
+                }
+
+                delete rules.max;
+                delete rules.min;
+
+                //we need to clear nodes that are not on the meta sheet
+                // and we have to initialize the counters
+                keys = Object.keys(rules);
+                for (i = 0; i < keys.length; i += 1) {
+                    if (!metaNodes[keys[i]]) {
+                        delete rules[keys[i]];
+                    } else {
+                        typeCounters[keys[i]] = 0;
+                    }
+                }
+
+                keys = Object.keys(rules);
+                for (i = 0; i < children.length; i += 1) {
+                    for (j = 0; j < keys.length; j += 1) {
+                        if (innerCore.isTypeOf(children[i], metaNodes[keys[j]])) {
+                            typeCounters[keys[j]] += 1;
+                        }
+                    }
+                }
+
+                i = validNodes.length;
+                keys = Object.keys(typeCounters);
+                while (i--) {
+                    for (j = 0; j < keys.length; j += 1) {
+                        if (rules[keys[j]].max &&
+                            rules[keys[j]].max > -1 &&
+                            rules[keys[j]].max <= typeCounters[keys[j]] &&
+                            innerCore.isTypeOf(validNodes[i], metaNodes[keys[j]])) {
+                            validNodes.splice(i, 1);
+                            break;
+                        }
+                    }
+                }
+            }
+
+            //before every next step we check if we still have potential nodes
+            if (validNodes.length === 0) {
+                return validNodes;
+            }
+
+            if (parameters.aspect) {
+                keys = innerCore.getAspectMeta(node, parameters.aspect);
+                i = validNodes.length;
+
+                while (i--) {
+                    inAspect = false;
+                    for (j = 0; j < keys.length; j += 1) {
+                        if (innerCore.isTypeOf(validNodes[i], metaNodes[keys[j]])) {
+                            inAspect = true;
+                            break;
+                        }
+                    }
+                    if (!inAspect) {
+                        validNodes.splice(i, 1);
```

Comment:
This is potentially slow, make a FIXME note.
http://jsperf.com/array-push-vs-splice